File: tests/zig/f2s.zig
const std = @import("std");
const profile = @import("profile.zig");

const total_numbers = 10000;
const total_times = 4;
var nums: [total_numbers]f32 = undefined;
var result: usize = 0;
var times: u32 = total_times;

fn mkFloat(i: usize) f32 {
    const f: f32 = @floatFromInt(i);
    return f + f / @as(f32, total_numbers);
}

noinline fn f2s(rounds: u32) usize {
    var sum: usize = 0;
    var buf: [64]u8 = undefined;
    for (0..rounds) |_| {
        for (nums) |f| {
            const s = std.fmt.bufPrint(&buf, "{d}", .{f}) catch unreachable;
            sum += s.len;
        }
    }
    return sum;
}

fn run(_: void) void {
    result = f2s(@as(*volatile u32, &times).*);
    std.mem.doNotOptimizeAway(result);
}

pub fn main(init: std.process.Init) !void {
    for (0..total_numbers) |i| nums[i] = mkFloat(i);
    try profile.profile(init.io, "float2string", run, {});
}

File: tests/cpp/f2s.cpp
#include "profile.h"
#include <cstring>
#include <vector>

static const int TOTAL_NUMBERS = 10000;
static volatile int times = 4;

static float mk_float(int i) { return float(i) + float(i) / float(TOTAL_NUMBERS); }

PROFILE_NOINLINE int32_t test_f2s(const std::vector<float> & nums, int rounds) {
    int32_t summ = 0;
    char buffer[128];
    for (int i = 0; i != rounds; ++i) {
        for (float f : nums) {
            snprintf(buffer, sizeof(buffer), "%f", f);
            summ += (int32_t) strlen(buffer);
        }
    }
    return summ;
}

int main() {
    std::vector<float> nums(TOTAL_NUMBERS);
    for (int i = 0; i < TOTAL_NUMBERS; i++) nums[i] = mk_float(i);
    int32_t result = 0;
    profile::run("float2string", [&] { result = test_f2s(nums, times); profile::keep(result); });
    return result > 0 ? 0 : 1;
}
